File: Sources/Reflection/Utilities/UpdateChecker.swift
import Foundation
import os.log

// MARK: - Types

struct UpdateInfo: Sendable {
    let latestVersion: String
    let currentVersion: String
    let releaseURL: URL
    let releaseNotes: String
}

// MARK: - URLDataProvider protocol (testability)

protocol URLDataProvider: Sendable {
    func data(from url: URL) async throws -> (Data, URLResponse)
}

extension URLSession: URLDataProvider {}

// MARK: - UpdateChecker

@MainActor
final class UpdateChecker {
    private static let logger = Logger(
        subsystem: Bundle.main.bundleIdentifier ?? "com.reflection.app",
        category: "UpdateChecker"
    )

    private static let repoOwner = "SatanshuMishra"
    private static let repoName = "reflection"
    private static let lastSeenVersionKey = "lastSeenVersion"

    private let dataProvider: URLDataProvider
    let currentVersion: String

    init(
        dataProvider: URLDataProvider = URLSession.shared,
        currentVersion: String? = nil
    ) {
        self.dataProvider = dataProvider
        self.currentVersion = currentVersion
            ?? Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
            ?? "0.0.0"
    }

    // MARK: - Public API

    /// Checks if a newer version is available on GitHub Releases.
    /// Returns nil if current version is up-to-date, or on any error.
    func checkForUpdate() async -> UpdateInfo? {
        let urlString = "https://api.github.com/repos/\(Self.repoOwner)/\(Self.repoName)/releases/latest"
        guard let url = URL(string: urlString) else {
            Self.logger.error("Invalid GitHub API URL")
            return nil
        }

        do {
            let (data, _) = try await dataProvider.data(from: url)
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let tagName = json["tag_name"] as? String,
                  let htmlURL = json["html_url"] as? String,
                  let releaseURL = URL(string: htmlURL) else {
                Self.logger.warning("Failed to parse GitHub release response")
                return nil
            }

            let remoteVersion = stripVersionPrefix(tagName)
            let releaseNotes = json["body"] as? String ?? ""

            guard isVersion(remoteVersion, newerThan: currentVersion) else {
                Self.logger.info("App is up to date (current: \(self.currentVersion), latest: \(remoteVersion))")
                return nil
            }

            Self.logger.info("Update available: \(remoteVersion) (current: \(self.currentVersion))")
            return UpdateInfo(
                latestVersion: remoteVersion,
                currentVersion: currentVersion,
                releaseURL: releaseURL,
                releaseNotes: releaseNotes
            )
        } catch {
            Self.logger.error("Failed to check for updates: \(error.localizedDescription)")
            return nil
        }
    }

    /// Fetches release notes for a specific version tag from GitHub.
    func fetchReleaseNotes(for version: String) async -> String? {
        let tag = version.hasPrefix("v") ? version : "v\(version)"
        let urlString = "https://api.github.com/repos/\(Self.repoOwner)/\(Self.repoName)/releases/tags/\(tag)"
        guard let url = URL(string: urlString) else { return nil }

        do {
            let (data, _) = try await dataProvider.data(from: url)
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let body = json["body"] as? String,
                  !body.isEmpty else {
                return nil
            }
            return body
        } catch {
            Self.logger.error("Failed to fetch release notes for \(tag): \(error.localizedDescription)")
            return nil
        }
    }

    // MARK: - What's New tracking

    static var lastSeenVersion: String? {
        get { UserDefaults.standard.string(forKey: lastSeenVersionKey) }
        set { UserDefaults.standard.set(newValue, forKey: lastSeenVersionKey) }
    }

    var isFirstLaunchOfNewVersion: Bool {
        guard let lastSeen = Self.lastSeenVersion else {
            return true // First ever launch
        }
        return lastSeen != currentVersion
    }

    // MARK: - Version comparison

    /// Returns true if `version` is strictly newer than `other`.
    func isVersion(_ version: String, newerThan other: String) -> Bool {
        let lhs = parseVersion(version)
        let rhs = parseVersion(other)

        for i in 0..<max(lhs.count, rhs.count) {
            let l = i < lhs.count ? lhs[i] : 0
            let r = i < rhs.count ? rhs[i] : 0
            if l > r { return true }
            if l < r { return false }
        }
        return false // equal
    }

    // MARK: - Private helpers

    private func stripVersionPrefix(_ tag: String) -> String {
        tag.hasPrefix("v") ? String(tag.dropFirst()) : tag
    }

    private func parseVersion(_ version: String) -> [Int] {
        version.split(separator: ".").compactMap { Int($0) }
    }
}

File: windows/src/utilities/UpdateChecker.cpp
#include "utilities/UpdateChecker.h"
#include "utilities/Logger.h"

#include <sstream>
#include <vector>

namespace reflection {

namespace {

/// Parse a version string like "1.2.3" into components.
std::vector<int> parse_version(const std::string& version) {
    std::vector<int> parts;
    std::string cleaned = version;

    // Strip leading 'v' or 'vw' prefix
    if (!cleaned.empty() && cleaned[0] == 'v') {
        cleaned = cleaned.substr(1);
    }
    if (!cleaned.empty() && cleaned[0] == 'w') {
        cleaned = cleaned.substr(1);
    }

    std::istringstream stream(cleaned);
    std::string segment;
    while (std::getline(stream, segment, '.')) {
        try {
            parts.push_back(std::stoi(segment));
        } catch (...) {
            parts.push_back(0);
        }
    }
    return parts;
}

} // namespace

std::optional<UpdateInfo> UpdateChecker::check(const std::string& current_version) {
    Logger::info("Checking for updates (current: {})", current_version);

    // TODO (Milestone 6): Fetch from GitHub API via WinHTTP
    // Parse JSON response for tag_name, html_url, body
    // Compare versions
    // Return UpdateInfo if newer version available

    return std::nullopt;
}

bool UpdateChecker::is_newer(const std::string& remote, const std::string& local) {
    const auto remote_parts = parse_version(remote);
    const auto local_parts = parse_version(local);

    const size_t max_len = std::max(remote_parts.size(), local_parts.size());

    for (size_t i = 0; i < max_len; ++i) {
        const int r = (i < remote_parts.size()) ? remote_parts[i] : 0;
        const int l = (i < local_parts.size()) ? local_parts[i] : 0;

        if (r > l) return true;
        if (r < l) return false;
    }

    return false; // Equal
}

} // namespace reflection
